plutus-ir/src/Language/PlutusIR/Compiler/Datatype.hs

```diff
@@ -153,142 +219,137 @@ mkDatatypePatternFunctor d@(Datatype _ _ tvs _ _) = do
+
+    -- constructor args and their types
+    argsAndTypes <- do
+        let argTypes = constructorArgTypes constr
+        -- we don't have any names for these things, we just had the type, so we call them "arg_i
+        argNames <- for [0..(length argTypes -1)] (\i -> liftQuote $ freshName () $ strToBs $ "arg_" ++ show i)
+        pure $ zip argNames argTypes
+
+    -- See Note [Recursive datatypes]
+    let maybeWrap t = case r of
+            Rec    -> PLC.Wrap () (tyVarDeclName tn) pf t
+            NonRec -> t
+
+    pure $
+        -- /\t_1 .. t_n
+        mkIterTyAbs (fmap splitTyVarDecl tvs) $
+        -- \arg_1 .. arg_m
+        mkIterLamAbs argsAndTypes $
+        -- wrap
+        maybeWrap $
         -- no need for a body value check here, we know it's a lambda (see Note [Value restriction])
-        -- forall r . cfuncs
-        resAbstracted = PLC.TyAbs () resultTypeName resultKind cfuncs
-        -- wrap resAbstracted
-        fixed = case r of
-            Rec    -> PLC.Wrap () n pf resAbstracted
-            NonRec -> resAbstracted
-        -- \a_1 .. a_m . fixed
-        afuncs = mkIterLamAbs argNamesAndTypes fixed
-    in afuncs
+        -- forall out
+        PLC.TyAbs () resultType (PLC.Type ()) $
+        -- \case_1 .. case_j
+        mkIterLamAbs casesAndTypes $
+        -- c_i arg_1 .. arg_m
+        mkIterApp thisCase (fmap (PLC.Var () . fst) argsAndTypes)
 
 -- Destructors
 
 -- See note [Scott encoding of datatypes]
--- See note [Abstract data types]
 -- | Make the destructor for a 'Datatype'.
+-- @
+--     mkDestructor List = /\(a :: *) . \(x : List a) . unwrap x
+-- @
 mkDestructor :: MonadQuote m => Recursivity -> Datatype TyName Name () -> m (PLC.Term TyName Name ())
 mkDestructor r (Datatype _ tn tvs _ _) =
     let
         tyVars = fmap splitTyVarDecl tvs
     in do
+        -- See note [Recursive datatypes]
+        let maybeUnwrap body = case r of
+                Rec    -> PLC.Unwrap () body
+                NonRec -> body
+
         -- t t_1 .. t_n
         let applied = mkIterTyApp (PLC.TyVar () $ tyVarDeclName tn) (fmap (PLC.TyVar () . fst) tyVars)
 
-        -- this is basically the identity! but we do need to unwrap it
         x <- liftQuote $ freshName () "x"
-        let body = PLC.Var () x
-            unwrapped = case r of
-                Rec    -> PLC.Unwrap () body
-                NonRec -> body
-            lambda = PLC.LamAbs () x applied unwrapped
-            -- /\ t_1 .. t_n
-            abstracted = mkIterTyAbs tyVars lambda
-
-        pure abstracted
+        pure $
+            -- /\t_1 .. t_n
+            mkIterTyAbs tyVars $
+            -- \x
+            PLC.LamAbs () x applied $
+            -- unwrap
+            maybeUnwrap $
+            PLC.Var () x
 
+-- See note [Scott encoding of datatypes]
 -- | Make the type of a destructor for a 'Datatype'.
+-- @
+--     mkDestructorTy
+--         (\(a :: *) -> forall (out_List :: *) . (r -> (a -> List a -> r) -> r))
+--         List =
```

Comment:
I again suggest to add the

```
mkDestructorTy ListF List =
mkDestructorTy  (\(a :: *) -> forall (out_List :: *) . (r -> (a -> List a -> r) -> r)) =
```

line. Also you bind `out_List`, but use `r`. (I like `r` more in snippets like that, because it's easier to observe an entire at once)
--         List =

Reply:
Yeah, I also like `r` but I'm trying to be consistent, and I think the more explicit naming helps that.